crypto/hpke/hpke.c

```diff
@@ -0,0 +1,2178 @@
+        erv = 0;
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    ctx->shared_secretlen = lsslen;
+    memcpy(ctx->shared_secret, lss, lsslen);
+
+err:
+    EVP_PKEY_CTX_free(pctx);
+    EVP_PKEY_free(spub);
+    OPENSSL_cleanse(lss, sizeof(lss));
+    return erv;
+}
+
+static int hpke_do_rest(OSSL_HPKE_CTX *ctx, int operation,
+                        unsigned char *ct, size_t *ctlen,
+                        unsigned char *pt, size_t *ptlen,
+                        const unsigned char *info, size_t infolen,
+                        const unsigned char *aad, size_t aadlen)
+{
+    int erv = 1;
+    size_t ks_contextlen = OSSL_HPKE_MAXSIZE;
+    unsigned char ks_context[OSSL_HPKE_MAXSIZE];
+    size_t halflen = 0;
+    size_t pskidlen = 0;
+    size_t psk_hashlen = OSSL_HPKE_MAXSIZE;
+    unsigned char psk_hash[OSSL_HPKE_MAXSIZE];
+    int kem_ind = 0, kdf_ind = 0, aead_ind = 0;
+    size_t secretlen = OSSL_HPKE_MAXSIZE;
+    unsigned char secret[OSSL_HPKE_MAXSIZE];
+    size_t noncelen = OSSL_HPKE_MAXSIZE;
+    unsigned char nonce[OSSL_HPKE_MAXSIZE];
+    unsigned char seqbuf[12];
+    size_t seqlen = 0;
+    size_t keylen = OSSL_HPKE_MAXSIZE;
+    unsigned char key[OSSL_HPKE_MAXSIZE];
+    EVP_KDF_CTX *kctx = NULL;
+    unsigned char suitebuf[6];
+    const char *mdname = NULL;
+
+    if ((kem_ind = kem_iana2index(ctx->suite.kem_id)) == 0) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    if ((aead_ind = aead_iana2index(ctx->suite.aead_id)) == 0) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    if ((kdf_ind = kdf_iana2index(ctx->suite.kdf_id)) == 0) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    mdname = hpke_kdf_tab[kdf_ind].mdname;
+    kctx = ossl_kdf_ctx_create("HKDF", mdname, ctx->libctx, ctx->propq);
+    if (kctx == NULL) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    /* create key schedule context */
+    memset(ks_context, 0, sizeof(ks_context));
+    ks_context[0] = (unsigned char)(ctx->mode % 256);
+    ks_contextlen--; /* remaining space */
+    halflen = hpke_kdf_tab[kdf_ind].Nh;
+    if ((2 * halflen) > ks_contextlen) {
+        erv = 0;
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    pskidlen = (ctx->psk == NULL ? 0 : strlen(ctx->pskid));
+    /* mode == FULL as per RFC9180 sec 5.1 */
+    suitebuf[0] = ctx->suite.kem_id / 256;
+    suitebuf[1] = ctx->suite.kem_id % 256;
+    suitebuf[2] = ctx->suite.kdf_id / 256;
+    suitebuf[3] = ctx->suite.kdf_id % 256;
+    suitebuf[4] = ctx->suite.aead_id / 256;
+    suitebuf[5] = ctx->suite.aead_id % 256;
+    erv = ossl_hpke_labeled_extract(kctx, ks_context + 1, halflen,
+                                    NULL, 0, OSSL_HPKE_SEC51LABEL,
+                                    suitebuf, 6, OSSL_HPKE_PSKIDHASH_LABEL,
```

Comment:
Magic number 6 should be sizeof(suitebuf) in multiple places..

Reply:
ack, done